### ISSUE TITLE
[css-grid-3][masonry] variable track sizes + dense packing has poor performance

### ISSUE THREAD

**Issue by fantasai** [w3c/csswg-drafts#9326]
From @bfgeek in https://github.com/w3c/csswg-drafts/issues/9041#issuecomment-1710806191
> Variable track sizes is problematic with items that span multiple tracks with "dense" packing. To correctly place the item in the correct tracks, you need O(N) layout passes on the item.

**Comment by tabatkins** [comment]
It also interacts badly if you want to allow "auto-spanning", where an element will get a span based on its width. This might be a good mode to allow, possibly even with some author control - if an image is >X% of the track width, make it span 2 instead of 1 (or a larger span, as appropriate).

With variable-width tracks this is *at minimum* problematic for layout, and might be inherently circular (I haven't given it enough thought yet to be certain).

**Comment by fantasai** [comment]
@tabatkins I think auto-spanning is a separate issue. :) And an equally applicable request for Grid.

**Comment by tabatkins** [comment]
Nah, it's part of this issue - *if* we do auto-spanning (and I think Masonry libraries do sometimes have this feature), then variable-size tracks are even more problematic. (And it would be equally problematic for Grid, I think.)

**Comment by fantasai** [comment]
@tabatkins I don't see what that has to do with dense packing. Filed as https://github.com/w3c/csswg-drafts/issues/10089

**Comment by fantasai** [comment]
Back to the original issue...

Agreed that having N different track sizes means dense packing would require O(N) layout passes on items added after a gap, and this would be a problem. We suggest adopting a heuristic that brings this closer to O(2) for most cases:

1. Lay out the item in the widest track, and in the narrowest track, and find the shortest height among the two.
2. Skip trying to fit into any gaps that are shorter than that height.

It's possible, but unlikely, that an item's actual shortest height would be at a middling width; and this allows us to avoid a lot of that extra work in the much more common cases where the item definitely won't fit.

**Comment by fantasai** [comment]
Ok, discussing with @stwrt, new proposal for dense-packing in masonry:
1. Lay out the item in the track where it would be without dense packing.
2. If the item is short enough to fit into an earlier gap _in a track with the same used size as this track_, move it into that gap.

This avoids doing more than one layout, and it works as expected for explicitly-placed masonry in mixed-size tracks as well as for auto-placed masonry in consistently-sized tracks, and does a reasonable job of dense-packing large collections of auto-placed masonry in mixed-size tracks as well.

**Comment by astearns** [comment]
In step 2 does it have to be exactly the same used size, or could it be “an earlier gap in a track with a used size at least as large as this track”?

**Comment by tabatkins** [comment]
Would have to be the same size to actually work; a larger track size *usually* would make the item shorter (and thus more likely to fit), but not *always* - say, an aspect-ratio image that fills the track. And even if it did fit, it would still be a second layout, potentially ~doubling the layout cost from turning on dense packing.

-------

This algo sounds reasonable to me? In the common case of equal-sized tracks, it'll work naturally to fill holes everywhere; in mixed tracks, it'll at least do *some* work, without increasing costs beyond that of tracking gaps (which is trivial).

**Comment by ethanjv** [comment]
The latest algorithm sounds reasonable, as Tab commented, the only cost of this algorithm is keeping track of the gaps. This is fairly doable if we store the current and all previous running positions for each track in the grid axis. However, when you consider items spanning multiple tracks there are a couple of issues that need clarifying:

1. Is it the expectation that, when looking for a previous gap for an item spanning multiple tracks, we use the maximum running position across spanned tracks? The figure below exemplifies how the item might be placed into multiple possible gaps.

![masonry-dense-packing-example](https://github.com/user-attachments/assets/7e31e04e-49e7-44aa-abec-b93600437760)

2. Combining gaps from multiple tracks is not an easy process, the algorithm would need to consider the complexity of matching a gap from one track with (possibly more than) one gap in the next tracks, implying non-trivial decision making.

Consider trying to place an item spanning 2 tracks in the example below (where all tracks have the same used size), choosing the upper gap in the third track produces the gap highlighted in pink:
![masonry-dense-packing-gap-00](https://github.com/user-attachments/assets/43c3d0cd-5857-4699-b4db-dd128ba331fa)

While choosing the lower gap produces the following smaller gap:
![masonry-dense-packing-gap-01](https://github.com/user-attachments/assets/f6a04b9f-5ec8-498e-a2a9-245768fe9db9)

**Comment by fantasai** [comment]
@ethanjv Your diagrams are amazing.

**Comment by tabatkins** [comment]
For @ethanjv's (1): yes, your displayed position options are exactly right.

For (2), we'd pick the higher area in the first picture.
